<h1 id="pipeline-for-recombination-suppression-and-selection-affect-local-ancestries-in-genomes-of-a-migratory-songbird">Pipeline for “Recombination suppression and selection affect local ancestries in genomes of a migratory songbird”</h1>
<h2 id="local-pca">Local PCA</h2>
<p>Here I show how local PCA was run with three example chromosomes, 12, 20, and 21, which are listed in <code>local_PCA/list/chromosomes/list</code></p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true"></a><span class="va">dirbase=$PWD</span>/local_PCA</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true"></a><span class="va">dirin=$dirbase</span>/input</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true"></a><span class="va">dirvcf=$dirin</span>/vcf</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true"></a><span class="va">dirtab=$dirin</span>/table</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true"></a><span class="va">dirlist=$dirbase</span>/list</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true"></a><span class="va">dirout=$dirbase</span>/output</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true"></a><span class="va">dirlocalpca=$dirout</span>/local_PCA</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true"></a><span class="va">dirlocalpcamds=$dirout</span>/local_PCA_MDS</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true"></a><span class="va">dirscripts=$dirbase</span>/scripts</span></code></pre></div>
<p>Download filtered VCF files for chromosomes 12, 20, 21.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true"></a></span></code></pre></div>
<p>Make list of sites from VCF.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true"></a><span class="kw">while</span> <span class="bu">read</span> <span class="va">chr</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true"></a><span class="kw">do</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true"></a>        <span class="ex">bcftools</span> query -f <span class="st">&#39;%POS\n&#39;</span> <span class="va">$dirvcf</span>/<span class="va">$chr</span>.vcf.gz <span class="op">&gt;</span> <span class="va">$dirtab</span>/<span class="va">$chr</span>.sites.list</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true"></a><span class="kw">done</span> <span class="op">&lt;</span><span class="va">$dirlist</span><span class="ex">/chromosomes.list</span></span></code></pre></div>
<p>Make genotype table from VCF.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true"></a></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true"></a><span class="kw">while</span> <span class="bu">read</span> <span class="va">chr</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true"></a><span class="kw">do</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true"></a>        <span class="ex">sbatch</span> <span class="va">$dirscripts</span>/vcf2table.sh <span class="va">$dirvcf</span>/<span class="va">$chr</span>.vcf.gz <span class="va">$dirtab</span>/<span class="va">$chr</span>.table</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true"></a><span class="kw">done</span> <span class="op">&lt;</span><span class="va">$dirlist</span><span class="ex">/chromosomes.list</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true"></a></span></code></pre></div>
<p>Run local PCA. <code>local_PCA/scripts/local_PCA.sh</code> submits <code>local_PCA/scripts/local_PCA.R</code> via slurm. Check <code>local_PCA/scripts/local_PCA.sh</code> and <code>local_PCA/scripts/local_PCA.R</code> for detail.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true"></a><span class="ex">module</span> load R/3.5.3</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true"></a></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true"></a><span class="kw">while</span> <span class="bu">read</span> <span class="va">chr</span> </span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true"></a><span class="kw">do</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true"></a>        <span class="ex">sbatch</span> <span class="va">$dirscripts</span>/local_PCA.sh <span class="va">$dirtab</span> <span class="va">$dirtab</span> <span class="va">$dirlocalpca</span> <span class="va">$dirlocalpcamds</span> <span class="va">$dirlist</span> <span class="va">$chr</span> <span class="va">$dirscripts</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true"></a><span class="kw">done</span> <span class="op">&lt;</span><span class="va">$dirlist</span><span class="ex">/chromosomes.list</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true"></a></span></code></pre></div>
<p>Concatenate the output for the three chromosomes.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true"></a></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true"></a><span class="kw">while</span> <span class="bu">read</span> <span class="va">chr</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true"></a><span class="kw">do</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true"></a>        <span class="fu">cat</span> <span class="va">$dirlocalpcamds</span>/local_PCA_MDS_<span class="va">$chr</span>.txt</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true"></a><span class="kw">done</span> <span class="op">&lt;</span><span class="va">$dirlist</span><span class="ex">/chromosomes.list</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;NR==1{print $0}NR&gt;1{if($1!=&quot;chr&quot;)print $0}&#39;</span> <span class="op">&gt;</span> <span class="va">$dirlocalpcamds</span>/local_PCA_MDS_3_chromosomes.txt</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true"></a></span></code></pre></div>
<p>Get threshold of MDS values and plot MDS distribution.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true"></a></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true"></a><span class="ex">Rscript</span> <span class="va">$dirscripts</span>/plot_local_PCA_MDS_distribution.R --input <span class="va">$dirlocalpcamds</span>/local_PCA_MDS_3_chromosomes.txt --output <span class="va">$dirlocalpcamds</span>/local_PCA_MDS_3_chromosomes --chrlist <span class="va">$dirlist</span>/chromosomes.list</span></code></pre></div>
<p><img src="local_PCA/output/local_PCA_MDS/local_PCA_MDS_3_chromosomes_MDS1_distribution.png" /></p>
<p><img src="local_PCA/output/local_PCA_MDS/local_PCA_MDS_3_chromosomes_MDS2_distribution.png" /></p>
<p>Plot MDS1 vs MDS2.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true"></a></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true"></a><span class="ex">Rscript</span> <span class="va">$dirscripts</span>/plot_local_PCA_MDS_MDS1-vs-MDS2.R --input <span class="va">$dirlocalpcamds</span>/local_PCA_MDS_3_chromosomes.txt --thre <span class="va">$dirlocalpcamds</span>/local_PCA_MDS_3_chromosomesthreshold.txt --output <span class="va">$dirlocalpcamds</span>/local_PCA_MDS --chrlist <span class="va">$dirlist</span>/chromosomes.list</span></code></pre></div>
<p><img src="local_PCA/output/local_PCA_MDS/local_PCA_MDS_MDS1-vs-MDS2_outliers.png" /></p>
<p>Get coordinates of outlier windows.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true"></a><span class="ex">Rscript</span> <span class="va">$dirscripts</span>/getOutliers.R --input <span class="va">$dirlocalpcamds</span>/local_PCA_MDS_3_chromosomes.txt --thre <span class="va">$dirlocalpcamds</span>/local_PCA_MDS_3_chromosomesthreshold.txt --output <span class="va">$dirlocalpcamds</span>/local_PCA_MDS --chrlist <span class="va">$dirlist</span>/chromosomes.list</span></code></pre></div>
<p>Make Manhattan plots.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true"></a><span class="ex">Rscript</span> <span class="va">$dirscripts</span>/plot_local_PCA_MDS_manhattan.R --input <span class="va">$dirlocalpcamds</span>/local_PCA_MDS_3_chromosomes.txt --thre <span class="va">$dirlocalpcamds</span>/local_PCA_MDS_3_chromosomesthreshold.txt --bed <span class="va">$dirlocalpcamds</span>/local_PCA_MDS_outlier.bed --output <span class="va">$dirlocalpcamds</span>/local_PCA_MDS --chrlist <span class="va">$dirlist</span>/chromosomes_length.list</span></code></pre></div>
<p><img src="local_PCA/output/local_PCA_MDS/local_PCA_MDS_MDS1_outliers.png" /></p>
<p><img src="local_PCA/output/local_PCA_MDS/local_PCA_MDS_MDS2_outliers.png" /></p>
<h2 id="pca-in-local-pca-outliers">PCA in local PCA outliers</h2>
<p>Here I show how PCA was run for three local PCA outlier regions of chromosomes 12, 20, and 21, which represent class-1, 2, and 3 outliers.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true"></a><span class="va">dirbase=$PWD</span>/PCA</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true"></a><span class="va">dirin=$dirbase</span>/input</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true"></a><span class="va">dirvcf=$dirin</span>/vcf</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true"></a><span class="va">dirlist=$dirbase</span>/list</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true"></a><span class="va">dirout=$dirbase</span>/output</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true"></a><span class="va">dirscripts=$dirbase</span>/scripts</span></code></pre></div>
<p>Link VCF files from <code>local_PCA/input/vcf</code> and index them.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true"></a><span class="fu">ln</span> <span class="va">$dirbase</span>/../local_PCA/input/vcf/*vcf.gz <span class="va">$dirvcf</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true"></a></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true"></a><span class="kw">while</span> <span class="bu">read</span> <span class="va">chr</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true"></a><span class="kw">do</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true"></a>        <span class="bu">echo</span> <span class="va">$chr</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true"></a>        <span class="ex">bcftools</span> index <span class="va">$dirvcf</span>/<span class="va">$chr</span>.vcf.gz</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true"></a><span class="kw">done</span><span class="op">&lt;</span><span class="va">$dirlist</span><span class="ex">/chromosomes.list</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true"></a></span></code></pre></div>
<p>Copy BED file for coordinates of local PCA outliers of chromosomes 12, 20, and 21.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true"></a><span class="fu">ln</span> <span class="va">$dirbase</span>/../local_PCA/output/local_PCA_MDS/local_PCA_MDS_outlier.bed <span class="va">$dirlist</span></span></code></pre></div>
<p>Extract SNPs within the local PCA outlier regions.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true"></a></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true"></a><span class="kw">while</span> <span class="bu">read</span> <span class="va">chr</span> <span class="va">from</span> <span class="va">to</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true"></a><span class="kw">do</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true"></a>        <span class="ex">bcftools</span> view -O z -r <span class="va">$chr</span>:<span class="va">$from</span>-<span class="va">$to</span> -S <span class="va">$dirlist</span>/blackcap_id.list <span class="va">$dirvcf</span>/<span class="va">$chr</span>.vcf.gz <span class="kw">|</span> <span class="ex">vcftools</span> --gzvcf - --max-missing 0.9 --recode --recode-INFO-all -c <span class="kw">|</span><span class="ex">bgzip</span> <span class="op">&gt;</span> <span class="va">$dirvcf</span>/<span class="va">${chr}</span>_<span class="va">${from}</span>_<span class="va">$to</span>.vcf.gz</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true"></a><span class="kw">done</span> <span class="op">&lt;</span> <span class="va">$dirlist</span><span class="ex">/local_PCA_MDS_outlier.bed</span></span></code></pre></div>
<p>Run <code>PLINK</code> for PCA.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true"></a></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true"></a><span class="kw">while</span> <span class="bu">read</span> <span class="va">chr</span> <span class="va">from</span> <span class="va">to</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true"></a><span class="kw">do</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true"></a>        <span class="ex">sbatch</span> <span class="va">$dirscripts</span>/plink_pca.sh  <span class="va">$dirvcf</span>/<span class="va">${chr}</span>_<span class="va">${from}</span>_<span class="va">$to</span> <span class="va">$dirout</span>/<span class="va">${chr}</span>_<span class="va">${from}</span>_<span class="va">$to</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true"></a><span class="kw">done</span> <span class="op">&lt;</span> <span class="va">$dirlist</span><span class="ex">/local_PCA_MDS_outlier.bed</span></span></code></pre></div>
<p>Plot PCA results.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true"></a></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true"></a><span class="ex">Rscript</span> <span class="va">$dirscripts</span>/plot_pca_per_outlier.R --dirpca <span class="va">$dirout</span> --poplist <span class="va">$dirlist</span>/id_spp_pop_site_pheno.tsv --outlierlist <span class="va">$dirlist</span>/local_PCA_MDS_outlier.bed --chrlist <span class="va">$dirlist</span>/chromosomes.list</span></code></pre></div>
<p><img src="PCA/output/PCA_localPCA_outlier.png" /></p>
<p>Get Eigenvalues.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true"></a></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true"></a><span class="kw">while</span> <span class="bu">read</span> <span class="va">chr</span> <span class="va">pos1</span> <span class="va">pos2</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true"></a><span class="kw">do</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true"></a>        <span class="fu">awk</span> -v chr=<span class="va">$chr</span> -v pos1=<span class="va">$pos1</span> -v pos2=<span class="va">$pos2</span> <span class="st">&#39;BEGIN{printf &quot;%s %s %s &quot;, chr,pos1,pos2}{printf &quot;%s &quot;,$1}END{print &quot;&quot;}&#39;</span> <span class="va">$dirout</span>/<span class="va">${chr}</span>_<span class="va">${pos1}</span>_<span class="va">${pos2}</span>.eigenval</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true"></a><span class="kw">done</span> <span class="op">&lt;</span> <span class="va">$dirlist</span><span class="ex">/local_PCA_MDS_outlier.bed</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;NR==1{printf &quot;%s %s %s &quot;,&quot;chr&quot;,&quot;pos1&quot;,&quot;pos2&quot;;for(i=4;i&lt;=NF;i++){printf &quot;%s &quot;,&quot;PC&quot;i-3};print &quot;&quot;}{print $0}&#39;</span><span class="op">&gt;</span> <span class="va">$dirout</span>/eigenvalues.txt</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true"></a></span></code></pre></div>
<p>Plot Eigenvalues.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true"></a></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true"></a><span class="ex">Rscript</span> <span class="va">$dirscripts</span>/plot_eigenval.R --dirpca <span class="va">$dirout</span> </span></code></pre></div>
<p><img src="PCA/output/eigenval.png" /></p>
<h2 id="class-1-genomic-islands">Class-1 genomic islands</h2>
<div class="sourceCode" id="cb19"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true"></a></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true"></a><span class="va">dirbase=$PWD</span>/class-1</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true"></a><span class="va">dirin=$dirbase</span>/input</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true"></a><span class="va">dirvcf=$dirin</span>/vcf</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true"></a><span class="va">dirlist=$dirbase</span>/list</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true"></a><span class="va">dirout=$dirbase</span>/output</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true"></a><span class="va">dirscripts=$dirbase</span>/scripts</span></code></pre></div>
<h3 id="heterozygosity">Heterozygosity</h3>
<p>Make link to VCF of chromosome 12, as an example chromosome harbouring a class-1 genomic island.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true"></a><span class="fu">ln</span> <span class="va">$dirbase</span>/../PCA/input/vcf/chr_12.vcf.gz  <span class="va">$dirvcf</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true"></a><span class="fu">ln</span> <span class="va">$dirbase</span>/../PCA/input/vcf/chr_12.vcf.gz.csi  <span class="va">$dirvcf</span></span></code></pre></div>
<p>Get coordinates of class-1 genomic island of chromosome 12.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true"></a></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true"></a><span class="fu">awk</span> -v OFS=<span class="st">&quot;\t&quot;</span> <span class="st">&#39;$1==&quot;chr_12&quot;{$1=$1;print $0}&#39;</span> <span class="va">$dirbase</span>/../local_PCA/output/local_PCA_MDS/local_PCA_MDS_outlier.bed <span class="op">&gt;</span> <span class="va">$dirlist</span>/class-1.chr_12.bed</span></code></pre></div>
<p>Make list of individuals with AA, AB, and BB.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true"></a></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true"></a><span class="fu">awk</span> <span class="st">&#39;$3&lt;0{print $1}&#39;</span> <span class="va">$dirbase</span>/../PCA/output/chr_12_14126710_22227355.eigenvec <span class="op">&gt;</span> <span class="va">$dirlist</span>/chr_12.AA.list</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true"></a><span class="fu">awk</span> <span class="st">&#39;$3&gt;0.15{print $1}&#39;</span> <span class="va">$dirbase</span>/../PCA/output/chr_12_14126710_22227355.eigenvec <span class="op">&gt;</span> <span class="va">$dirlist</span>/chr_12.BB.list</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true"></a><span class="fu">awk</span> <span class="st">&#39;$3&gt;0&amp;&amp;$3&lt;0.15{print $1}&#39;</span> <span class="va">$dirbase</span>/../PCA/output/chr_12_14126710_22227355.eigenvec <span class="op">&gt;</span> <span class="va">$dirlist</span>/chr_12.AB.list</span></code></pre></div>
<p>Compute heterozygosity.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true"></a></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true"></a><span class="kw">while</span> <span class="bu">read</span> <span class="va">chr</span> <span class="va">pos1</span> <span class="va">pos2</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true"></a><span class="kw">do</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true"></a>        <span class="kw">for</span> <span class="ex">geno</span> in AA AB BB</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true"></a>        <span class="kw">do</span></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true"></a>                <span class="kw">while</span> <span class="bu">read</span> <span class="va">id</span></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true"></a>                <span class="kw">do</span></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true"></a>                        <span class="co"># echo $id $chr $geno</span></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true"></a>                        <span class="ex">bcftools</span> query -f <span class="st">&#39;[%GT ]\n&#39;</span> -r <span class="va">$chr</span>:<span class="va">$pos1</span>-<span class="va">$pos2</span> -s <span class="va">$id</span> <span class="va">$dirvcf</span>/<span class="va">$chr</span>.vcf.gz <span class="kw">|</span><span class="fu">sed</span> <span class="st">&#39;s@0/0@0@g;s@0|0@0@g;s@1/1@0@g;s@1|1@0@g;s@0/1@1@g;s@0|1@1@g;s@1|0@1@g;s@\./\.@@g;s@\.|\.@@g&#39;</span> <span class="kw">|</span> <span class="fu">awk</span> -v chr=<span class="va">$chr</span> -v geno=<span class="va">$geno</span> -v id=<span class="va">$id</span> <span class="st">&#39;{i++;s+=$1}END{print chr,geno,id,s,i,s/i}&#39;</span></span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true"></a>                <span class="kw">done</span><span class="op">&lt;</span><span class="va">$dirlist</span><span class="ex">/</span><span class="va">${chr}</span><span class="ex">.</span><span class="va">$geno</span>.list</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true"></a>        <span class="kw">done</span> </span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true"></a><span class="kw">done</span> <span class="op">&lt;</span> <span class="va">$dirlist</span><span class="ex">/class-1.chr_12.bed</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;BEGIN{print &quot;chr&quot;,&quot;geno&quot;,&quot;id&quot;,&quot;n.het&quot;,&quot;n.sites&quot;,&quot;het&quot;}{print $0}&#39;</span> <span class="op">&gt;</span> <span class="va">$dirout</span>/chr_12.het.txt</span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true"></a></span></code></pre></div>
<p>Plot heterozygosity for AA, AB, and BB.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true"></a></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true"></a><span class="ex">Rscript</span> <span class="va">$dirscripts</span>/plot_het.R --dir <span class="va">$dirout</span></span></code></pre></div>
<p><img src="class-1/output/chr_12.het.png" /></p>
<h3 id="fst-dxy-and-π">F<sub>ST</sub>, d<sub>XY</sub> and π</h3>
<p><span class="math inline"><em>F</em><sub><em>S</em><em>T</em></sub></span> <span class="math inline"><em>π</em></span></p>
